src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs

```diff
@@ -1262,275 +1296,1280 @@ protected override NestedResourceInfoScope CreateNestedResourceInfoScope(WriterS
-        {
             this.jsonLightResourceSerializer.JsonLightValueSerializer.AssertRecursionDepthIsZero();
             if (resource.NonComputedProperties != null)
             {
-                this.jsonLightResourceSerializer.WriteProperties(
+                await this.jsonLightResourceSerializer.WritePropertiesAsync(
                     this.ResourceType,
                     resource.NonComputedProperties,
                     false /* isComplexValue */,
                     this.DuplicatePropertyNameChecker,
-                    resource.MetadataBuilder);
+                    resource.MetadataBuilder).ConfigureAwait(false);
                 this.jsonLightResourceSerializer.JsonLightValueSerializer.AssertRecursionDepthIsZero();
             }
         }
 
         /// <summary>
-        /// Writes the reason annotation for a delta deleted resource.
+        /// Asynchronously finish writing a resource.
         /// </summary>
-        /// <param name="resource">The resource to write the reason for.</param>
-        private void WriteDeltaResourceReason(ODataDeletedResource resource)
+        /// <param name="resource">The resource to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        protected override async Task EndResourceAsync(ODataResource resource)
         {
-            Debug.Assert(resource != null, "resource != null");
-
-            if (!resource.Reason.HasValue)
+            if (resource == null)
             {
                 return;
             }
 
-            this.jsonWriter.WriteName(JsonLightConstants.ODataReasonPropertyName);
+            // Get the projected properties
+            JsonLightResourceScope resourceScope = this.CurrentResourceScope;
 
-            switch (resource.Reason.Value)
+            await this.jsonLightResourceSerializer.WriteResourceMetadataPropertiesAsync(resourceScope)
+                .ConfigureAwait(false);
+
+            // Write custom instance annotations
+            await this.instanceAnnotationWriter.WriteInstanceAnnotationsAsync(
+                resource.InstanceAnnotations,
+                resourceScope.InstanceAnnotationWriteTracker).ConfigureAwait(false);
+
+            await this.jsonLightResourceSerializer.WriteResourceEndMetadataPropertiesAsync(
+                resourceScope,
+                resourceScope.DuplicatePropertyNameChecker).ConfigureAwait(false);
+
+            // Close the object scope
+            await this.asynchronousJsonWriter.EndObjectScopeAsync()
+                .ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Asynchronously start writing a property.
+        /// </summary>
+        /// <param name="property">The property info to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        protected override Task StartPropertyAsync(ODataPropertyInfo property)

```

Comment:
Since we defer to `ODataJsonLightPropertySerializer`'s `WritePropertyInfoAsync` method, property is null scenario would be caught by the call to `ValidatePropertyInfo` method.